tests/scripts/audit-validity-dates.py

```diff
@@ -0,0 +1,427 @@
+            line = self.buf[self._line_start:]
+            self._line_start = self.buf_len
+            self._line_no += 1
+
+        # Convert byte array to string with correct encoding and
+        # strip any whitespaces added in the decoding process.
+        return line.decode(sys.getdefaultencoding()).rstrip() + '\n'
+
+    def get_line_no(self):
+        """
+        Gives current line number.
+        """
+        return self._line_no
+
+    line_no = property(get_line_no)
+
+class SuiteDataAuditor(Auditor):
+    """Class for auditing files in tests/suites/*.data"""
+    def __init__(self, options):
+        super().__init__(options)
+        self.default_files = self.collect_default_files()
+
+    def collect_default_files(self):
+        """Collect all files in tests/suites/*.data"""
+        test_dir = self.find_test_dir()
+        suites_data_folder = os.path.join(test_dir, 'suites')
+        data_files = glob.glob(os.path.join(suites_data_folder, '*.data'))
+        return data_files
+
+    def parse_file(self, filename: str):
+        """
+        Parse a list of AuditData from file.
+
+        :param filename: name of the file to parse.
+        :return list of AuditData parsed from the file.
+        """
+        audit_data_list = []
+        data_f = FileWrapper(filename)
+        for _, _, _, test_args in parse_suite_data(data_f):
+            for idx, test_arg in enumerate(test_args):
+                match = re.match(r'"(?P<data>[0-9a-fA-F]+)"', test_arg)
+                if not match:
+                    continue
+                if not X509Parser.check_hex_string(match.group('data')):
+                    continue
+                audit_data = self.parse_bytes(bytes.fromhex(match.group('data')))
+                if audit_data is None:
+                    continue
+                audit_data.filename = "{}:{}:{}".format(filename,
+                                                        data_f.line_no,
+                                                        idx + 1)
+                audit_data_list.append(audit_data)
+
+        return audit_data_list
+
+def list_all(audit_data: AuditData):
+    print("{}\t{}\t{}\t{}".format(
+        audit_data.not_valid_before.isoformat(timespec='seconds'),
+        audit_data.not_valid_after.isoformat(timespec='seconds'),
+        audit_data.data_type.name,
+        audit_data.filename))
+
+def main():
+    """
+    Perform argument parsing.
+    """
+    parser = argparse.ArgumentParser(description=__doc__)
+
+    parser.add_argument('-a', '--all',
+                        action='store_true',
+                        help='list the information of all the files')
+    parser.add_argument('-v', '--verbose',
+                        action='store_true', dest='verbose',
+                        help='show warnings')
+    parser.add_argument('--not-before', dest='not_before',
+                        help=('not valid before this date (UTC, YYYY-MM-DD). '
```

Comment:
Does this mean “the not-valid-before date of the object is this date”, “the object is not valid before this date” (i.e. exclude objects whose validity starts before this date), or “the object's not-valid-before date is before this date” (i.e. exclude objects whose validity starts after this date)?

Reply:
The _behavior_ makes sense to me. But the _description_ of the option doesn't. Taken literally, it seems to say “only print objects whose not-before date is exactly this”. Or maybe “only print objects that are never valid before this”. Or maybe “only print objects such that there is a date before this when they're invalid”.

What the option actually does is “only print objects whose not-before date is after this date”, but that's not what the description says.

Suggested descriptions, taking each option independently:

> Validity starts before this date (i.e. not-before is after this date)

and

> Validity ends after this date (i.e. not-after is before this date)

But it probably makes more sense to take the options together and invert their meaning. The script lists “bad” objects. The good objects are the ones that are valid all the time from date1 to date2. So I like your proposal of `--from` and `--to` as option names, and then they would have descriptions referring to the good validity period. Something like:

> (general description) Exclude objects that are valid throughout the desired validity period.

> `--from DATE` Start of desired validity period (default: today).

> `--to DATE` End of desired validity period (default: --from).